File: ios/streaming/cache/HlsCacheStore.swift
import CommonCrypto
import Foundation
import OSLog

private let cacheLogger = Logger(subsystem: "com.nitroplay.hls", category: "CacheStore")

final class HlsCacheStore {
  private var maxBytes: Int
  private let ttlSeconds: TimeInterval = 7 * 24 * 60 * 60
  private let cacheDir: URL
  private let indexUrl: URL
  private var index: [String: HlsCacheEntry] = [:]
  private let queue = DispatchQueue(label: "hls-cache-store")
  private var pendingSave: DispatchWorkItem?

  init(maxBytes: Int = HlsCacheBudget.defaultMaxBytes) {
    self.maxBytes = HlsCacheBudget.normalize(maxBytes)
    let base = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
    cacheDir = base.appendingPathComponent("hls-cache", isDirectory: true)
    indexUrl = cacheDir.appendingPathComponent("index.json")
    loadIndex()
    ensureDir()
  }

  func setMaxBytes(_ bytes: Int) {
    queue.sync {
      maxBytes = HlsCacheBudget.normalize(bytes)
      evictIfNeeded()
      saveIndex()
    }
  }

  func get(url: String) -> Data? {
    guard let entry = queue.sync(execute: { index[url] }) else { return nil }
    if isExpired(entry) || !isSafeFileName(entry.fileName) {
      queue.async { [weak self] in
        self?.remove(url: url)
      }
      return nil
    }

    let fileUrl = cacheDir.appendingPathComponent(entry.fileName)
    guard let data = try? Data(contentsOf: fileUrl),
          data.count > 0,
          data.count == entry.size
    else {
      queue.async { [weak self] in
        self?.remove(url: url)
      }
      return nil
    }

    queue.async { [weak self] in
      guard var current = self?.index[url] else { return }
      current.lastAccess = Date().timeIntervalSince1970
      self?.index[url] = current
      self?.scheduleSave()
    }

    return data
  }

  func put(url: String, data: Data, streamKey: String?) {
    queue.async {
      self.ensureDir()
      self.evictIfNeeded()
      let name = self.sha256(url) + ".seg"
      let fileUrl = self.cacheDir.appendingPathComponent(name)
      do {
        try data.write(to: fileUrl, options: .atomic)
        let entry = HlsCacheEntry(
          url: url,
          fileName: name,
          size: data.count,
          streamKey: streamKey,
          createdAt: Date().timeIntervalSince1970,
          lastAccess: Date().timeIntervalSince1970
        )
        self.index[url] = entry
        self.evictIfNeeded()
        self.scheduleSave()
      } catch {
        cacheLogger.error("Failed to write cache entry for \(url): \(error.localizedDescription)")
        return
      }
    }
  }

  func has(url: String) -> Bool {
    queue.sync {
      guard let entry = index[url] else { return false }
      if isExpired(entry) || !isSafeFileName(entry.fileName) {
        remove(url: url)
        return false
      }
      let fileUrl = cacheDir.appendingPathComponent(entry.fileName)
      guard let size = fileSize(fileUrl), size > 0, size == entry.size else {
        remove(url: url)
        return false
      }
      return true
    }
  }

  func getFilePath(url: String) -> URL? {
    queue.sync {
      guard let entry = index[url] else { return nil }
      if isExpired(entry) || !isSafeFileName(entry.fileName) {
        remove(url: url)
        return nil
      }
      let fileUrl = cacheDir.appendingPathComponent(entry.fileName)
      guard let size = fileSize(fileUrl), size > 0, size == entry.size else {
        remove(url: url)
        return nil
      }
      var current = entry
      current.lastAccess = Date().timeIntervalSince1970
      index[url] = current
      scheduleSave()
      return fileUrl
    }
  }

  func getCacheStats() -> [String: Any] {
    queue.sync {
      evictExpired()
      let totalSize = index.values.reduce(0) { $0 + $1.size }
      return [
        "totalSize": totalSize,
        "fileCount": index.count,
        "maxSize": maxBytes
      ]
    }
  }

  func getCacheStats(streamKey: String) -> [String: Any] {
    queue.sync {
      evictExpired()
      let totalSize = index.values.reduce(0) { $0 + $1.size }
      let streamEntries = index.values.filter { $0.streamKey == streamKey }
      return [
        "totalSize": totalSize,
        "fileCount": index.count,
        "maxSize": maxBytes,
        "streamSize": streamEntries.reduce(0) { $0 + $1.size },
        "streamFileCount": streamEntries.count
      ]
    }
  }

  func clearAll() {
    queue.sync {
      for entry in self.index.values {
        let fileUrl = self.cacheDir.appendingPathComponent(entry.fileName)
        try? FileManager.default.removeItem(at: fileUrl)
      }
      self.index.removeAll()
      // Sweep any orphan files (written but never indexed, or left by a crash) so
      // the directory matches the now-empty index.
      let urls = (try? FileManager.default.contentsOfDirectory(at: self.cacheDir, includingPropertiesForKeys: nil)) ?? []
      for url in urls where url.lastPathComponent != self.indexUrl.lastPathComponent {
        try? FileManager.default.removeItem(at: url)
      }
      self.saveIndex()
    }
  }

  func clearThumbnails() {
    queue.async {
      // Drop indexed thumbnails (deletes file + index entry) ...
      let indexedThumbs = self.index.values.filter { $0.fileName.hasSuffix(".thumb") }
      for entry in indexedThumbs {
        self.remove(url: entry.url)
      }
      // ... and sweep any legacy/un-indexed thumbnail files left on disk.
      let urls = (try? FileManager.default.contentsOfDirectory(at: self.cacheDir, includingPropertiesForKeys: nil)) ?? []
      urls
        .filter { $0.lastPathComponent.hasSuffix(".thumb") }
        .forEach { try? FileManager.default.removeItem(at: $0) }
      // Reconcile: drop any index entry whose backing file no longer exists.
      for entry in Array(self.index.values)
      where !FileManager.default.fileExists(atPath: self.cacheDir.appendingPathComponent(entry.fileName).path) {
        self.index.removeValue(forKey: entry.url)
      }
      // Persist immediately so the on-disk index never lags the cleared state.
      self.saveIndex()
    }
  }

  private func ensureDir() {
    if !FileManager.default.fileExists(atPath: cacheDir.path) {
      try? FileManager.default.createDirectory(at: cacheDir, withIntermediateDirectories: true)
    }
  }

  private func evictIfNeeded() {
    evictExpired()
    var total = index.values.reduce(0) { $0 + $1.size }
    guard total > maxBytes else { return }
    let target = HlsCacheBudget.evictionTarget(for: maxBytes)

    var streams: [String: [HlsCacheEntry]] = [:]
    for entry in index.values {
      let key = entry.streamKey ?? entry.url
      streams[key, default: []].append(entry)
    }
    let sorted = streams.sorted { stream1, stream2 in
      let oldest1 = stream1.value.map(\.lastAccess).min() ?? 0
      let oldest2 = stream2.value.map(\.lastAccess).min() ?? 0
      return oldest1 < oldest2
    }

    for (_, entries) in sorted {
      if total <= target { break }
      for entry in entries {
        remove(url: entry.url)
        total -= entry.size
      }
    }
  }

  private func evictExpired() {
    let now = Date().timeIntervalSince1970
    let entries = Array(index.values)
    for entry in entries where now - entry.createdAt > ttlSeconds {
      remove(url: entry.url)
    }
  }

  private func remove(url: String) {
    guard let entry = index[url] else { return }
    let fileUrl = cacheDir.appendingPathComponent(entry.fileName)
    try? FileManager.default.removeItem(at: fileUrl)
    index.removeValue(forKey: url)
    scheduleSave()
  }

  private func isExpired(_ entry: HlsCacheEntry) -> Bool {
    let now = Date().timeIntervalSince1970
    return now - entry.createdAt > ttlSeconds
  }

  private func loadIndex() {
    guard let data = try? Data(contentsOf: indexUrl) else { return }
    if let decoded = try? JSONDecoder().decode([String: HlsCacheEntry].self, from: data) {
      index = decoded.filter { _, entry in
        guard isSafeFileName(entry.fileName),
              let size = fileSize(cacheDir.appendingPathComponent(entry.fileName))
        else {
          return false
        }
        return size > 0 && size == entry.size
      }
    } else {
      index = [:]
      try? FileManager.default.removeItem(at: indexUrl)
    }
  }

  private func scheduleSave() {
    pendingSave?.cancel()
    let work = DispatchWorkItem { [weak self] in
      self?.saveIndex()
    }
    pendingSave = work
    queue.asyncAfter(deadline: .now() + 5, execute: work)
  }

  private func saveIndex() {
    if let data = try? JSONEncoder().encode(index) {
      try? data.write(to: indexUrl, options: .atomic)
    }
  }

  private func sha256(_ input: String) -> String {
    let data = Data(input.utf8)
    var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    data.withUnsafeBytes {
      _ = CC_SHA256($0.baseAddress, CC_LONG(data.count), &hash)
    }
    return hash.map { String(format: "%02x", $0) }.joined()
  }

  private func isSafeFileName(_ fileName: String) -> Bool {
    guard !fileName.isEmpty,
          fileName == (fileName as NSString).lastPathComponent,
          !fileName.contains("/"),
          !fileName.contains("\\"),
          !fileName.contains("..")
    else {
      return false
    }
    return true
  }

  private func fileSize(_ url: URL) -> Int? {
    guard let attributes = try? FileManager.default.attributesOfItem(atPath: url.path),
          let size = attributes[.size] as? NSNumber
    else {
      return nil
    }
    return size.intValue
  }

  // Thumbnails share the segment index so they participate in the same TTL and
  // size eviction. They are keyed under a "thumb:" namespace to avoid colliding
  // with a segment cached under the same URL; the on-disk file keeps the plain
  // sha256(url).thumb name so previously written thumbnails still resolve.
  private func thumbnailKey(_ url: String) -> String { "thumb:\(url)" }

  func putThumbnail(url: String, data: Data) -> URL? {
    queue.sync {
      ensureDir()
      evictIfNeeded()
      let name = sha256(url) + ".thumb"
      let fileUrl = cacheDir.appendingPathComponent(name)
      do {
        try data.write(to: fileUrl, options: .atomic)
        let now = Date().timeIntervalSince1970
        let key = thumbnailKey(url)
        index[key] = HlsCacheEntry(
          url: key,
          fileName: name,
          size: data.count,
          streamKey: nil,
          createdAt: now,
          lastAccess: now
        )
        evictIfNeeded()
        scheduleSave()
        return fileUrl
      } catch {
        cacheLogger.error("Failed to write thumbnail for \(url): \(error.localizedDescription)")
        return nil
      }
    }
  }

  func getThumbnailPath(url: String) -> URL? {
    queue.sync {
      let name = sha256(url) + ".thumb"
      let fileUrl = cacheDir.appendingPathComponent(name)
      let key = thumbnailKey(url)
      if let entry = index[key] {
        if isExpired(entry) || !FileManager.default.fileExists(atPath: fileUrl.path) {
          remove(url: key)
          return nil
        }
        var current = entry
        current.lastAccess = Date().timeIntervalSince1970
        index[key] = current
        scheduleSave()
        return fileUrl
      }
      // Legacy thumbnail written before indexing: register it lazily so it now
      // participates in TTL/size eviction.
      guard FileManager.default.fileExists(atPath: fileUrl.path) else { return nil }
      let attrs = try? FileManager.default.attributesOfItem(atPath: fileUrl.path)
      let size = (attrs?[.size] as? Int) ?? 0
      let now = Date().timeIntervalSince1970
      index[key] = HlsCacheEntry(
        url: key,
        fileName: name,
        size: size,
        streamKey: nil,
        createdAt: now,
        lastAccess: now
      )
      scheduleSave()
      return fileUrl
    }
  }

  func hasThumbnail(url: String) -> Bool {
    queue.sync {
      let key = thumbnailKey(url)
      if let entry = index[key], isExpired(entry) {
        remove(url: key)
        return false
      }
      let name = sha256(url) + ".thumb"
      let fileUrl = cacheDir.appendingPathComponent(name)
      return FileManager.default.fileExists(atPath: fileUrl.path)
    }
  }
}

struct HlsCacheEntry: Codable {
  let url: String
  let fileName: String
  let size: Int
  let streamKey: String?
  let createdAt: TimeInterval
  var lastAccess: TimeInterval
}

File: android/src/main/java/com/nitroplay/streaming/cache/HlsCacheStore.kt
package com.nitroplay.video.streaming.cache

import android.content.Context
import android.net.Uri
import org.json.JSONObject
import java.io.File
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

class HlsCacheStore(context: Context, maxBytes: Long = HlsCacheBudget.DEFAULT_MAX_BYTES) {
    @Volatile
    private var maxBytes: Long = HlsCacheBudget.normalize(maxBytes)
    private val ttlMs: Long = 7L * 24L * 60L * 60L * 1000L
    private val cacheDir = File(context.cacheDir, "hls-cache")
    private val indexFile = File(cacheDir, "index.json")
    private val index = ConcurrentHashMap<String, HlsCacheEntry>()
    private val saveExecutor = Executors.newSingleThreadScheduledExecutor()
    private var pendingSave: ScheduledFuture<*>? = null

    init {
        if (!cacheDir.exists()) cacheDir.mkdirs()
        loadIndex()
    }

    fun setMaxBytes(bytes: Long) {
        maxBytes = HlsCacheBudget.normalize(bytes)
        evictIfNeeded()
        saveIndex()
    }

    fun has(url: String): Boolean {
        val entry = index[url] ?: return false
        if (isExpired(entry) || !isSafeFileName(entry.fileName)) {
            remove(url)
            return false
        }
        val file = File(cacheDir, entry.fileName)
        if (!file.exists() || file.length() <= 0L || file.length() != entry.size) {
            remove(url)
            return false
        }
        return true
    }

    fun getFilePath(url: String): File? {
        val entry = index[url] ?: return null
        if (isExpired(entry) || !isSafeFileName(entry.fileName)) {
            remove(url)
            return null
        }
        val file = File(cacheDir, entry.fileName)
        if (!file.exists() || file.length() <= 0L || file.length() != entry.size) {
            remove(url)
            return null
        }
        entry.lastAccess = System.currentTimeMillis()
        scheduleSave()
        return file
    }

    fun get(url: String): ByteArray? {
        val entry = index[url] ?: return null
        if (isExpired(entry) || !isSafeFileName(entry.fileName)) {
            remove(url)
            return null
        }
        val file = File(cacheDir, entry.fileName)
        if (!file.exists() || file.length() <= 0L || file.length() != entry.size) {
            remove(url)
            return null
        }
        entry.lastAccess = System.currentTimeMillis()
        scheduleSave()
        return file.readBytes()
    }

    fun put(url: String, data: ByteArray, streamKey: String?) {
        if (!cacheDir.exists()) cacheDir.mkdirs()
        evictIfNeeded()
        val name = sha256(url) + ".seg"
        val file = File(cacheDir, name)
        file.writeBytes(data)
        index[url] = HlsCacheEntry(url, name, data.size.toLong(), streamKey, System.currentTimeMillis(), System.currentTimeMillis())
        evictIfNeeded()
        scheduleSave()
    }

    fun getCacheStats(): Map<String, Any> {
        evictExpired()
        return mapOf(
            "totalSize" to index.values.sumOf { it.size },
            "fileCount" to index.size,
            "maxSize" to maxBytes
        )
    }

    fun getStreamCacheStats(streamKey: String): Map<String, Any> {
        evictExpired()
        val streamEntries = index.values.filter { it.streamKey == streamKey }
        return mapOf(
            "totalSize" to index.values.sumOf { it.size },
            "fileCount" to index.size,
            "maxSize" to maxBytes,
            "streamSize" to streamEntries.sumOf { it.size },
            "streamFileCount" to streamEntries.size
        )
    }

    fun clearAll() {
        index.values.forEach { entry ->
            val file = File(cacheDir, entry.fileName)
            if (file.exists()) file.delete()
        }
        index.clear()
        // Sweep any orphan files (written but never indexed, or left by a crash) so
        // the directory matches the now-empty index.
        cacheDir
            .listFiles()
            ?.filter { it.name != indexFile.name }
            ?.forEach { it.delete() }
        saveIndex()
    }

    fun clearThumbnails() {
        // Drop indexed thumbnails (deletes file + index entry) ...
        index.values.filter { it.fileName.endsWith(".thumb") }.forEach { remove(it.url) }
        // ... and sweep any legacy/un-indexed thumbnail files left on disk.
        cacheDir
            .listFiles()
            ?.filter { it.name.endsWith(".thumb") }
            ?.forEach { it.delete() }
        // Reconcile: drop any index entry whose backing file no longer exists.
        index.values
            .filter { !File(cacheDir, it.fileName).exists() }
            .forEach { index.remove(it.url) }
        // Persist immediately so the on-disk index never lags the cleared state.
        saveIndex()
    }

    fun close() {
        pendingSave?.cancel(false)
        pendingSave = null
        saveIndex()
        saveExecutor.shutdownNow()
    }

    private fun evictIfNeeded() {
        evictExpired()
        var total = index.values.sumOf { it.size }
        if (total <= maxBytes) return
        val target = HlsCacheBudget.evictionTarget(maxBytes)

        val streams = index.values.groupBy { it.streamKey ?: it.url }
        val sorted = streams.entries.sortedBy { entry ->
            entry.value.minOfOrNull { it.lastAccess } ?: 0L
        }

        for ((_, entries) in sorted) {
            if (total <= target) break
            for (entry in entries) {
                total -= entry.size
                remove(entry.url)
            }
        }
    }

    private fun evictExpired() {
        val now = System.currentTimeMillis()
        index.values.filter { now - it.createdAt > ttlMs }.forEach { remove(it.url) }
    }

    private fun remove(url: String) {
        val entry = index[url] ?: return
        val file = File(cacheDir, entry.fileName)
        if (file.exists()) file.delete()
        index.remove(url)
        scheduleSave()
    }

    private fun isExpired(entry: HlsCacheEntry): Boolean {
        return System.currentTimeMillis() - entry.createdAt > ttlMs
    }

    private fun loadIndex() {
        if (!indexFile.exists()) return
        try {
            val text = indexFile.readText()
            if (text.isBlank()) return
            val json = JSONObject(text)
            json.keys().forEach { key ->
                val obj = json.getJSONObject(key)
                val entry = HlsCacheEntry(
                    url = key,
                    fileName = obj.getString("fileName"),
                    size = obj.getLong("size"),
                    streamKey = if (obj.has("streamKey") && !obj.isNull("streamKey")) obj.getString("streamKey") else null,
                    createdAt = obj.getLong("createdAt"),
                    lastAccess = obj.getLong("lastAccess")
                )
                val file = File(cacheDir, entry.fileName)
                if (isSafeFileName(entry.fileName) && file.exists() && file.length() > 0L && file.length() == entry.size) {
                    index[key] = entry
                }
            }
        } catch (_: Exception) {
            index.clear()
            indexFile.delete()
        }
    }

    private fun scheduleSave() {
        pendingSave?.cancel(false)
        pendingSave = saveExecutor.schedule({
            saveIndex()
        }, 5, TimeUnit.SECONDS)
    }

    @Synchronized
    private fun saveIndex() {
        val json = JSONObject()
        index.forEach { (url, entry) ->
            val obj = JSONObject()
            obj.put("fileName", entry.fileName)
            obj.put("size", entry.size)
            obj.put("streamKey", entry.streamKey)
            obj.put("createdAt", entry.createdAt)
            obj.put("lastAccess", entry.lastAccess)
            json.put(url, obj)
        }
        indexFile.writeText(json.toString())
    }

    private fun sha256(input: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(input.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }

    private fun isSafeFileName(fileName: String): Boolean {
        return fileName.isNotBlank() &&
            fileName == File(fileName).name &&
            !fileName.contains("/") &&
            !fileName.contains("\\") &&
            !fileName.contains("..")
    }

    // Thumbnails share the segment index so they participate in the same TTL and
    // size eviction. They are keyed under a "thumb:" namespace to avoid colliding
    // with a segment cached under the same URL; the on-disk file keeps the plain
    // sha256(url).thumb name so previously written thumbnails still resolve.
    private fun thumbnailKey(url: String): String = "thumb:$url"

    fun putThumbnail(url: String, data: ByteArray): String? {
        if (!cacheDir.exists()) cacheDir.mkdirs()
        evictIfNeeded()
        val name = "${sha256(url)}.thumb"
        val file = File(cacheDir, name)
        return try {
            file.writeBytes(data)
            val now = System.currentTimeMillis()
            val key = thumbnailKey(url)
            index[key] = HlsCacheEntry(key, name, data.size.toLong(), null, now, now)
            evictIfNeeded()
            scheduleSave()
            Uri.fromFile(file).toString()
        } catch (_: Exception) {
            null
        }
    }

    fun getThumbnailPath(url: String): String? {
        val name = "${sha256(url)}.thumb"
        val file = File(cacheDir, name)
        val key = thumbnailKey(url)
        val entry = index[key]
        if (entry != null) {
            if (isExpired(entry) || !file.exists()) {
                remove(key)
                return null
            }
            entry.lastAccess = System.currentTimeMillis()
            scheduleSave()
            return Uri.fromFile(file).toString()
        }
        // Legacy thumbnail written before indexing: register it lazily so it now
        // participates in TTL/size eviction.
        if (!file.exists()) return null
        val now = System.currentTimeMillis()
        index[key] = HlsCacheEntry(key, name, file.length(), null, now, now)
        scheduleSave()
        return Uri.fromFile(file).toString()
    }

    fun hasThumbnail(url: String): Boolean {
        val key = thumbnailKey(url)
        val entry = index[key]
        if (entry != null && isExpired(entry)) {
            remove(key)
            return false
        }
        val name = "${sha256(url)}.thumb"
        return File(cacheDir, name).exists()
    }
}

data class HlsCacheEntry(
    val url: String,
    val fileName: String,
    val size: Long,
    val streamKey: String?,
    val createdAt: Long,
    var lastAccess: Long
)
